arci-ros2/src/ros2_control.rs

```diff
@@ -0,0 +1,185 @@
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+
+use arci::*;
+use futures::stream::StreamExt;
+use parking_lot::Mutex;
+use r2r::{
+    builtin_interfaces::{msg as builtin_msg, msg::Time},
+    control_msgs::{action::FollowJointTrajectory, msg::JointTrajectoryControllerState},
+    std_msgs::msg::Header,
+    trajectory_msgs::msg as trajectory_msg,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::utils;
+
+/// Implement arci::JointTrajectoryClient for ROS2
+pub struct Ros2ControlClient {
+    state_topic: String,
+    action_client: r2r::ActionClient<FollowJointTrajectory::Action>,
+    /// r2r::Node to handle the action
+    node: Arc<Mutex<r2r::Node>>,
+    joint_names: Vec<String>,
+}
+
+// TODO:
+unsafe impl Sync for Ros2ControlClient {}
+
+impl Ros2ControlClient {
+    /// Creates a new `Ros2ControlClient` from ROS2 context and names of action and joints.
+    #[track_caller]
+    pub fn new(ctx: r2r::Context, action_name: &str, joint_names: Vec<String>) -> Self {
+        // TODO: Consider using unique name
+        let node = r2r::Node::create(ctx, "ros2_control_node", "arci_ros2").unwrap();
+        Self::from_node(node, action_name, joint_names)
+    }
+
+    /// Creates a new `Ros2ControlClient` from ROS2 node and names of action and joints.
+    #[track_caller]
+    pub fn from_node(mut node: r2r::Node, action_name: &str, joint_names: Vec<String>) -> Self {
+        let action_client = node
+            .create_action_client::<FollowJointTrajectory::Action>(&format!(
+                "{}/follow_joint_trajectory",
+                action_name
+            ))
+            .unwrap();
+        Self {
+            state_topic: format!("{}/state", action_name),
+            action_client,
+            node: Arc::new(Mutex::new(node)),
+            joint_names,
+        }
+    }
+}
+
+impl JointTrajectoryClient for Ros2ControlClient {
+    fn joint_names(&self) -> Vec<String> {
+        self.joint_names.clone()
+    }
+
+    fn current_joint_positions(&self) -> Result<Vec<f64>, arci::Error> {
+        let node = self.node.clone();
+        let mut state_subscriber = node
+            .lock()
+            .subscribe::<JointTrajectoryControllerState>(&self.state_topic)
+            .unwrap();
```

Comment:
It seems that subscribe should start in new() or something. It's too slow to do this.
It takes 1 sec for example.

Reply:
> It seems that subscribe should start in new() or something.

I originally wanted to do that, but I couldn't find a way to do it without "channel is full" warnings/errors.

> It takes 1 sec for example.

In my environment, it seems to be only at the first subscribe.

```
[arci-ros2/examples/ros2_control.rs:16] client.current_joint_positions() = Ok(
    [
        0.0,
        0.0,
    ],
)
[arci-ros2/examples/ros2_control.rs:17] i.elapsed() = 1.283533378s
[arci-ros2/examples/ros2_control.rs:25] i.elapsed() = 5.014140098s
[arci-ros2/examples/ros2_control.rs:27] client.current_joint_positions() = Ok(
    [
        0.8001994934739625,
        0.8001994934739625,
    ],
)
[arci-ros2/examples/ros2_control.rs:28] i.elapsed() = 13.427898ms
[arci-ros2/examples/ros2_control.rs:36] i.elapsed() = 5.014776349s
[arci-ros2/examples/ros2_control.rs:38] client.current_joint_positions() = Ok(
    [
        0.5730538883660978,
        0.5730538883660978,
    ],
)
[arci-ros2/examples/ros2_control.rs:39] i.elapsed() = 13.040018ms
```

So, it seems that the performance requirements can be met by doing the first subscribe at `from_node`.

```
[arci-ros2/examples/ros2_control.rs:16] client.current_joint_positions() = Ok(
    [
        0.0,
        0.0,
    ],
)
[arci-ros2/examples/ros2_control.rs:17] i.elapsed() = 84.003628ms
[arci-ros2/examples/ros2_control.rs:25] i.elapsed() = 5.012585296s
[arci-ros2/examples/ros2_control.rs:27] client.current_joint_positions() = Ok(
    [
        0.7994219067267148,
        0.7994219067267148,
    ],
)
[arci-ros2/examples/ros2_control.rs:28] i.elapsed() = 12.672687ms
[arci-ros2/examples/ros2_control.rs:36] i.elapsed() = 5.019069117s
[arci-ros2/examples/ros2_control.rs:38] client.current_joint_positions() = Ok(
    [
        0.5726392646129658,
        0.5726392646129658,
    ],
)
[arci-ros2/examples/ros2_control.rs:39] i.elapsed() = 10.538187ms
```

Reply:
> channel is full

Could you add some more information about the above warnings?

Reply:
> > channel is full
> 
> Could you add some more information about the above warnings?

Depending on the implementation, I mainly got either one of the following, or a more unfriendly error message which means the same thing (something like `error TrySendError { kind: Full }`).

```text
warning: could not send feedback msg (send failed because channel is full)
```

```text
warning: could not send feedback msg (send failed because receiver is gone)
```

(Warnings are occurring [here](https://github.com/sequenceplanner/r2r/blob/1e0a368d33c1ea56265af8bea1bc5f98804c22a2/src/action_clients.rs#L377), and [TrySendError is from futures](https://github.com/rust-lang/futures-rs/blob/dacf2565d4adfa978b342770190b9a6a1c56329a/futures-channel/src/mpsc/mod.rs#L216-L224))